weasyl/test/login/test_authenticate_bcrypt.py

```diff
@@ -0,0 +1,126 @@
+# encoding: utf-8
+"""
+Test suite for: login.py::def authenticate_bcrypt(username, password, session=True):
+"""
+import json
+
+from weasyl.test import db_utils
+from weasyl import login, macro
+from weasyl import define as d
+
+
+# Main test password
+raw_password = "0123456789"
+
+
+class Bag(object):
+    def __init__(self, **kw):
+        for kv in kw.items():
+            setattr(self, *kv)
+
+
+def test_logins_fail_if_no_username_provided():
+    result = login.authenticate_bcrypt(username='', password='password')
+    assert result == (0, 'invalid')
+
+
+def test_logins_fail_if_no_password_provided():
+    user_name = "test"
+    result = login.authenticate_bcrypt(username=user_name, password='')
+    assert result == (0, 'invalid')
+
+
+def test_login_fails_if_incorrect_username_is_provided():
+    user_name = "test"
+    result = login.authenticate_bcrypt(username=user_name, password=raw_password)
+    assert result == (0, 'invalid')
+
+
+def test_login_fails_for_incorrect_credentials():
+    user_name = "testAuthBcry0001"
```

Comment:
When users are actually created in the DB via ``user_id = db_utils.create_user(...)``, the username must be unique per constraints on the ``login`` table.
```
>       cursor.execute(statement, parameters)
E       IntegrityError: (psycopg2.IntegrityError) duplicate key value violates unique constraint "login_login_name_key"
E       DETAIL:  Key (login_name)=(test) already exists.
E        [SQL: 'INSERT INTO login (login_name, last_login) VALUES (%(login_name)s, %(last_login)s) RETURNING login.userid'] [parameters: {'last_login': -18000, 'login_name': u'test'}]
```

I made usernames unique where possible (read that as: where users aren't created, or where I don't need the username and so I just use ``db_utils.create_user(...)`` to auto-generate the username which it does (Line 30: https://github.com/Weasyl/weasyl/blob/master/weasyl/test/db_utils.py#L30 ). The alternative is to manually fetch the username which ``db_utils`` assigns, but again, I was advised to use the ``username=`` parameter for this purpose.

https://paste.weasyldev.com/show/8iqPkXKT9oVsrd7vD8lb/